### PR TITLE
chore: Improves the Select component UI/UX - iteration 4

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
Improves the Select component UI/UX - iteration 4.
- Migrates the native filters Select components to the new Select
- Fixes some bugs of the Select component related to searching and selection
- Simplifies the Select API for the async non-paginated version

Touched components:
```
src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx
src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx
src/filters/components/common.ts
src/filters/components/GroupBy/GroupByFilterPlugin.tsx
src/filters/components/Select/SelectFilterPlugin.tsx
src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx
src/filters/components/TimeGrain/TimeGrainFilterPlugin.tsx
```
@rusackas @villebro 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/124596960-40682000-de39-11eb-8ca5-0cb26af8080b.mp4

### TESTING INSTRUCTIONS
- Enter in a dashboard
- Open the filter bar
- Open the native filter modal
- Check the selects of the native filter modal
- After adding some filters check the selects in the filter bar

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by michael-s-molina** [comment]
@jinghua-qa @junlincc Can you help me test the native filters with the new select?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts
Comment (geido) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts:109]: Small typo here :)

**Review by geido — COMMENTED** [review]
The code is looking good! Just looking for the CI to pass and for the _env with native filters ON_ for some manual testing

**Comment by rusackas** [comment]
Hmmm... something seems amiss, if I'm not mistaken. If I go to the Sales Dashboard, and create a new filter, the default dataset should be "cleaned_sales_data" but the Select now shows "27"
image

"cleaned_sales_data" doesn't show up in the options list, but I _can_ search for it. When I select it, and create a filter, like so...
image

I get this error in the filter bar:
image

**Comment by rusackas** [comment]
Well hmm... on my local build, with the Sales Dashboard, I see the correct data source in the Select. However, I can't create functional native filters (attempted value and numeric range) on that dashboard. On other dashboards, things are working seemingly fine. Something is amiss, but I'm not yet sure if it's the data source, the dashboard, or the feature(s).

**Comment by michael-s-molina** [comment]
@rusackas Thanks for testing this. I'll investigate the problem with the Sales Dashboard.

**Comment by michael-s-molina** [comment]
@rusackas 

> Hmmm... something seems amiss, if I'm not mistaken. If I go to the Sales Dashboard, and create a new filter, the default dataset should be "cleaned_sales_data" but the Select now shows "27"

Fixed

> I get this error in the filter bar:
> image

When I try to add a default value I get "Error: Time column "OrderDate" does not exist in dataset". So probably this dataset is corrupted and you can ignore this error.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:67]: Teensy nit, but this could go above the `src/**/*` imports, with the other npm pkgs.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts
Comment (rusackas) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts:107]: Your logic here is all good... but for sport, this got me thinking... If you want to get fancy, you might be able to one-liner this with lodash magic... something like:
```
const mostUsedDataset = _(charts)
  .countBy('formData.datasource')
  .toPairs()
  .orderBy([1], ['desc'])
  .value().shift()[0];

return datasets[mostUsedDataset]?.id;
 ```
... but that's probably even more difficult to read 😜

Comment (michael-s-molina) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts:107]: Really fancy! Cool functions from Lodash! I think readability is more important though. I had to open Lodash docs to understand how it worked 🤣 .

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx
Comment (michael-s-molina) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx]: Done.

**Comment by michael-s-molina** [comment]
@rusackas I want to do a small change before merging. I'll request your review again when it's done.

**Comment by michael-s-molina** [comment]
@rusackas You can continue your review 😉

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:714]:
```suggestion
                    // We need to reset the column when the dataset has changed
```

File: superset-frontend/src/filters/components/GroupBy/GroupByFilterPlugin.tsx
Comment (rusackas) [inline superset-frontend/src/filters/components/GroupBy/GroupByFilterPlugin.tsx:23]: Hmm... this "Styles" could probably use a more specific name.

File: superset-frontend/src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx
Comment (rusackas) [inline superset-frontend/src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx:29]: If you agreed with the first one, this `Styles` could also use a more specific/explicit name as well. But then again, these are nits, and I'd love to see this merged!

**Review by rusackas — APPROVED** [review]
LGTM, but it'd be nice to have more eyes on this, as it's chunky!

**Comment by michael-s-molina** [comment]
@rusackas I'll do the suggested changes in a follow-up PR. @geido Can we get your approval as well?

**Comment by geido** [comment]
Sure @michael-s-molina . I'll do some manual testing as well today and come back to you.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx:131]: I think we could use the localization here

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DatasetSelect.tsx
Comment (geido) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DatasetSelect.tsx:112]: I think localizing the `aria-label `would be good. I see other places as well where it is not being localized. I won't mention them all. Those are easy to spot.

**Review by geido — APPROVED** [review]
The latest changes to the code look good. Just a nit below. Now going for manual testing

**Review by geido — CHANGES_REQUESTED** [review]
Hey @michael-s-molina please find the results of my manual testing efforts below:

I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.

Screen Shot 2021-07-19 at 15 25 47

I don't see my newly created option "test" in the dropdown in multiple mode

Screen Shot 2021-07-19 at 15 35 09

I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list

Screen Shot 2021-07-19 at 15 41 50


I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.

Screen Shot 2021-07-19 at 15 39 38

After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)

Screen Shot 2021-07-19 at 15 29 27

**Comment by michael-s-molina** [comment]
> Hey @michael-s-molina please find the results of my manual testing efforts below:
> 
> I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.
> 
> Screen Shot 2021-07-19 at 15 25 47
> 
> I don't see my newly created option "test" in the dropdown in multiple mode
> 
> Screen Shot 2021-07-19 at 15 35 09
> 
> I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list
> 
> Screen Shot 2021-07-19 at 15 41 50
> 
> I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.
> 
> Screen Shot 2021-07-19 at 15 39 38
> 
> After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)
> 
> Screen Shot 2021-07-19 at 15 29 27

Thanks for the throughout testing. None of the issues are related to this PR. The first 3 are related to the Select component and will be fixed in separate PRs. The last 2 are isolated issues. Will open separate PRs also. So the plan is to open these separate PRs and then rebase this PR on top of the fixes. So let's leave this open for now.

File: superset-frontend/src/filters/components/GroupBy/GroupByFilterPlugin.tsx
Comment (michael-s-molina) [reply superset-frontend/src/filters/components/GroupBy/GroupByFilterPlugin.tsx]: Done

File: superset-frontend/src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx
Comment (michael-s-molina) [reply superset-frontend/src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx]: Done

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx
Comment (michael-s-molina) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx]: Done

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DatasetSelect.tsx
Comment (michael-s-molina) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DatasetSelect.tsx]: Done

**Comment by michael-s-molina** [comment]
@geido @rusackas @junlincc 

> I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.
> 
> I don't see my newly created option "test" in the dropdown in multiple mode
> 
> I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list
> 

After investigation, I discovered that these were not bugs in the new `Select` component but problems in the `SelectFilterPlugin` component.  `SelectFilterPlugin` contained duplicated logic already present in the `Select` component. I removed the duplicated logic.

> I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.
> 

This was in fact related to this PR. Fixed.

> After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)
> 

This one is by design, until now we only keep the values if the user clicks on "Apply". We can change this behavior if the product team agrees but if that's the case we need to open another low priority issue.

@rusackas @geido I also applied all your suggestions.

@geido Thanks for testing the features. I updated the environment and requested your review again.

**Comment by geido** [comment]
Hey @michael-s-molina. Thanks for the fixes. There's only one problem that I could find. See below

When I choose _Time Grain_ and a default value, it won't allow me to select the option _Original value_. It fails to select it. This happens on both the modal and the left sidebar

Screen Shot 2021-07-20 at 17 12 03

Screen Shot 2021-07-20 at 17 24 03

**Comment by michael-s-molina** [comment]
@geido This problem is already happening on master so I'll open another PR to fix it so we can merge this one.

**Review by geido — APPROVED** [review]

**Pull request merged**
